# Проект от компании "Elama"

## Оглавление 

* [Библиотеки и описание таблиц](#id_1)    
* [Загрузка и слияние таблиц](#id_2)  
** [Сохранение данных](#id_3)  
** [Загрузка объединенного файла с таблицами](#id_4)  
* [Ответы на вопросы](#id_5)     
** [Вопрос 1](#id_6)   
** [Вопрос 2](#id_7)  
** [Вопрос 3](#id_8)  
** [Вопрос 4](#id_9)  
** [Вопрос 5](#id_10)   
* [Выводы](#id_11)  

### Библиотеки и описание таблиц <a class="anchor" id="id_1"></a>

In [17]:
import pandas as pd
from sqlalchemy import create_engine

Таблица events:  
a. id - идентификатор конкретного факта наступления события  
b. forecastMarker - идентификатор задачи прогнозирования (так мы называем
общую задачу прогноза конкретного бизнес-показателя. Есть отдельная
таблица, где хранятся свойства задач прогнозирования (источники данных,
гиперпараметры, регрессоры и т.д.), (необязательное поле)  
c. request_time - время когда зафиксировано событие  
d. event_id - идентификатор события  
e. sender - Имя системы поставившей задачу прогнозирования 
(необязательное поле)  
f. context - значения переменных или каких-либо параметров, связанных с
событием (необязательно поле)  
g. session_id - идентификатор исполнения конкретной задачи
прогнозирования (используется если задача пришла без forecastMarker и
sender). 
  
Таблица event_name - таблица с именами событий, которые случаются в сервисе:  
a. id - уникальный идентификатор  
b. event_name - название события в рамках сервиса прогнозирования


## Загрузка и слияние таблиц <a class="anchor" id="id_2"></a>

In [18]:
filename_1 = r'Elama\logs_2022-09-12.db'
filename_2 = r'Elama\logs_2022-09-13.db'
filename_3 = r'Elama\logs_2022-09-14.db'
merged_file =  r'Elama\merged_file.db'
engine = create_engine(f'sqlite:///{filename_1}', echo = False)

query = """
SELECT * 
FROM event_name
"""

event_name_12 = pd.read_sql(query, engine)
display(event_name_12)

id                                         event_name
0    1                     Инициализируем свойства класса
1    2        Переход в функцию getParametrsAndRegressors
2    3  Выход из функции getParametrsAndRegressors с п...
3    4  Выход из функции getParametrsAndRegressors БЕЗ...
4    5  Записали в свойства класса гиперпараметры и ре...
5    6  Начинаем читать данные из BQ по полученному SQ...
6    7   Получили данные из BQ по полученному SQL запросу
7    8                saveForecastRequest: вход в функцию
8    9              saveForecastRequest: выход из функции
9   10                             getDOW: вход в функцию
10  11                           getDOW: выход из функции
11  12                         getCutoffs: вход в функцию
12  13                       getCutoffs: выход из функции
13  14                       get_forecast: вход в функцию
14  15                     get_forecast: выход из функции
15  16               updateForecastAnswer: вход в функцию
16  17             updateForecastAnswer: выход из функции
17  18                      Инициализируем модель Prophet
18  19                 Обучаем модель на имеющихся данных
19  20                  Выполняем прогноз временного ряда
20  21                get_error_for_model: вход в функцию
21  22              get_error_for_model: выход из функции
22  23                  saveForecastToCSV: вход в функцию
23  24                saveForecastToCSV: выход из функции
24  25            getAdditionalRegressors: вход в функцию
25  26          getAdditionalRegressors: выход из функции
26  27       getForecastMarkerAsRegressor: вход в функцию
27  28     getForecastMarkerAsRegressor: выход из функции
28  29      Начинаем получать гипермараметры и регрессоры
29  30            Пустой набор данных для прогнозирования
30  31                             JSON отправлен обратно
31  32      get_error_for_model: импортировали библиотеки
32  33       get_error_for_model: прошли cross_validation
33  34  get_error_for_model: получили performance_metrics

In [19]:
query = """
SELECT * 
FROM events
"""

events_12 = pd.read_sql(query, engine)
display(events_12)

id                     forecastMarker                request_time  \
0          1           SelfService/RUS/turnover  2022-09-12 11:30:11.106628   
1          2           SelfService/RUS/turnover  2022-09-12 11:30:11.111359   
2          3           SelfService/RUS/turnover  2022-09-12 11:30:11.115989   
3          4           SelfService/RUS/turnover  2022-09-12 11:30:12.761754   
4          5           SelfService/RUS/turnover  2022-09-12 11:30:12.770134   
...      ...                                ...                         ...   
11202  11203  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.360209   
11203  11204  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.367637   
11204  11205  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.377117   
11205  11206  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.382090   
11206  11207  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.389464   

       event_id            sender  \
0             1       test_system   
1            29       test_system   
2             2       test_system   
3             3       test_system   
4             5       test_system   
...         ...               ...   
11202        24  weekly_report.ss   
11203        15  weekly_report.ss   
11204        16  weekly_report.ss   
11205        17  weekly_report.ss   
11206        31  weekly_report.ss   

                                                 context  \
0      df = <class 'NoneType'>, sql_query = <class 'N...   
1                                                   None   
2                                                   None   
3                          Найдено 1 строк с параметрами   
4                                                   None   
...                                                  ...   
11202                                               None   
11203                                               None   
11204                                               None   
11205                                               None   
11206                          Отдан датафрейм: (894, 4)   

                             session_id  
0      6f4277f6df57c4f4e90779966522c876  
1      6f4277f6df57c4f4e90779966522c876  
2      6f4277f6df57c4f4e90779966522c876  
3      6f4277f6df57c4f4e90779966522c876  
4      6f4277f6df57c4f4e90779966522c876  
...                                 ...  
11202  0d2086d69468e0587d193654a3a0b746  
11203  0d2086d69468e0587d193654a3a0b746  
11204  0d2086d69468e0587d193654a3a0b746  
11205  0d2086d69468e0587d193654a3a0b746  
11206  0d2086d69468e0587d193654a3a0b746  

[11207 rows x 7 columns]

In [20]:
engine = create_engine(f'sqlite:///{filename_2}', echo = False)
query = """
SELECT * 
FROM event_name
"""

event_name_13 = pd.read_sql(query, engine)
display(event_name_13)

id                                         event_name
0    1                     Инициализируем свойства класса
1    2        Переход в функцию getParametrsAndRegressors
2    3  Выход из функции getParametrsAndRegressors с п...
3    4  Выход из функции getParametrsAndRegressors БЕЗ...
4    5  Записали в свойства класса гиперпараметры и ре...
5    6  Начинаем читать данные из BQ по полученному SQ...
6    7   Получили данные из BQ по полученному SQL запросу
7    8                saveForecastRequest: вход в функцию
8    9              saveForecastRequest: выход из функции
9   10                             getDOW: вход в функцию
10  11                           getDOW: выход из функции
11  12                         getCutoffs: вход в функцию
12  13                       getCutoffs: выход из функции
13  14                       get_forecast: вход в функцию
14  15                     get_forecast: выход из функции
15  16               updateForecastAnswer: вход в функцию
16  17             updateForecastAnswer: выход из функции
17  18                      Инициализируем модель Prophet
18  19                 Обучаем модель на имеющихся данных
19  20                  Выполняем прогноз временного ряда
20  21                get_error_for_model: вход в функцию
21  22              get_error_for_model: выход из функции
22  23                  saveForecastToCSV: вход в функцию
23  24                saveForecastToCSV: выход из функции
24  25            getAdditionalRegressors: вход в функцию
25  26          getAdditionalRegressors: выход из функции
26  27       getForecastMarkerAsRegressor: вход в функцию
27  28     getForecastMarkerAsRegressor: выход из функции
28  29      Начинаем получать гипермараметры и регрессоры
29  30            Пустой набор данных для прогнозирования
30  31                             JSON отправлен обратно
31  32      get_error_for_model: импортировали библиотеки
32  33       get_error_for_model: прошли cross_validation
33  34  get_error_for_model: получили performance_metrics

In [21]:
query = """
SELECT * 
FROM events
"""

events_13 = pd.read_sql(query, engine)
display(events_13)

id                     forecastMarker                request_time  \
0          1                               None  2022-09-13 00:10:27.967745   
1          2                               None  2022-09-13 00:10:27.972425   
2          3                               None  2022-09-13 00:10:27.977883   
3          4                               None  2022-09-13 00:10:27.982433   
4          5                               None  2022-09-13 00:10:27.989409   
...      ...                                ...                         ...   
22270  22271  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.052806   
22271  22272  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.058017   
22272  22273  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.065892   
22273  22274  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.071837   
22274  22275  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.078643   

       event_id                 sender  \
0             1                   None   
1            29                   None   
2             5                   None   
3             8                   None   
4             9                   None   
...         ...                    ...   
22270        24  test.weekly_report.ss   
22271        15  test.weekly_report.ss   
22272        16  test.weekly_report.ss   
22273        17  test.weekly_report.ss   
22274        31  test.weekly_report.ss   

                                                 context  \
0      df = <class 'pandas.core.frame.DataFrame'>, sq...   
1                                                   None   
2                                                   None   
3                                                   None   
4                                                   None   
...                                                  ...   
22270                                               None   
22271                                               None   
22272                                               None   
22273                                               None   
22274                          Отдан датафрейм: (895, 4)   

                             session_id  
0      fa89a4d46b9b34761a580f0152f94f05  
1      fa89a4d46b9b34761a580f0152f94f05  
2      fa89a4d46b9b34761a580f0152f94f05  
3      fa89a4d46b9b34761a580f0152f94f05  
4      fa89a4d46b9b34761a580f0152f94f05  
...                                 ...  
22270  d637d1b6d2b70fd088e2ea44959483cb  
22271  d637d1b6d2b70fd088e2ea44959483cb  
22272  d637d1b6d2b70fd088e2ea44959483cb  
22273  d637d1b6d2b70fd088e2ea44959483cb  
22274  d637d1b6d2b70fd088e2ea44959483cb  

[22275 rows x 7 columns]

In [22]:
engine = create_engine(f'sqlite:///{filename_3}', echo = False)
query = """
SELECT * 
FROM event_name
"""

event_name_14 = pd.read_sql(query, engine)
display(event_name_14)

id                                         event_name
0    1                     Инициализируем свойства класса
1    2        Переход в функцию getParametrsAndRegressors
2    3  Выход из функции getParametrsAndRegressors с п...
3    4  Выход из функции getParametrsAndRegressors БЕЗ...
4    5  Записали в свойства класса гиперпараметры и ре...
5    6  Начинаем читать данные из BQ по полученному SQ...
6    7   Получили данные из BQ по полученному SQL запросу
7    8                saveForecastRequest: вход в функцию
8    9              saveForecastRequest: выход из функции
9   10                             getDOW: вход в функцию
10  11                           getDOW: выход из функции
11  12                         getCutoffs: вход в функцию
12  13                       getCutoffs: выход из функции
13  14                       get_forecast: вход в функцию
14  15                     get_forecast: выход из функции
15  16               updateForecastAnswer: вход в функцию
16  17             updateForecastAnswer: выход из функции
17  18                      Инициализируем модель Prophet
18  19                 Обучаем модель на имеющихся данных
19  20                  Выполняем прогноз временного ряда
20  21                get_error_for_model: вход в функцию
21  22              get_error_for_model: выход из функции
22  23                  saveForecastToCSV: вход в функцию
23  24                saveForecastToCSV: выход из функции
24  25            getAdditionalRegressors: вход в функцию
25  26          getAdditionalRegressors: выход из функции
26  27       getForecastMarkerAsRegressor: вход в функцию
27  28     getForecastMarkerAsRegressor: выход из функции
28  29      Начинаем получать гипермараметры и регрессоры
29  30            Пустой набор данных для прогнозирования
30  31                             JSON отправлен обратно
31  32      get_error_for_model: импортировали библиотеки
32  33       get_error_for_model: прошли cross_validation
33  34  get_error_for_model: получили performance_metrics

In [23]:
query = """
SELECT * 
FROM events
"""

events_14 = pd.read_sql(query, engine)
display(events_14)

id                         forecastMarker  \
0          1                                   None   
1          2                                   None   
2          3                                   None   
3          4                                   None   
4          5                                   None   
...      ...                                    ...   
11244  11245  weekly_report.SS:new_users_90d:tiktok   
11245  11246  weekly_report.SS:new_users_90d:tiktok   
11246  11247  weekly_report.SS:new_users_90d:tiktok   
11247  11248  weekly_report.SS:new_users_90d:tiktok   
11248  11249  weekly_report.SS:new_users_90d:tiktok   

                     request_time  event_id            sender  \
0      2022-09-14 00:09:39.234361         1              None   
1      2022-09-14 00:09:39.238851        29              None   
2      2022-09-14 00:09:39.242626         5              None   
3      2022-09-14 00:09:39.248606         8              None   
4      2022-09-14 00:09:39.254990         9              None   
...                           ...       ...               ...   
11244  2022-09-14 06:07:47.582138        24  weekly_report.ss   
11245  2022-09-14 06:07:47.588185        15  weekly_report.ss   
11246  2022-09-14 06:07:47.594110        16  weekly_report.ss   
11247  2022-09-14 06:07:47.601623        17  weekly_report.ss   
11248  2022-09-14 06:07:47.609898        31  weekly_report.ss   

                                                 context  \
0      df = <class 'pandas.core.frame.DataFrame'>, sq...   
1                                                   None   
2                                                   None   
3                                                   None   
4                                                   None   
...                                                  ...   
11244                                               None   
11245                                               None   
11246                                               None   
11247                                               None   
11248                         Отдан датафрейм: (1352, 4)   

                             session_id  
0      ec6a82e4a1b9b4dc467639eb1a5d6987  
1      ec6a82e4a1b9b4dc467639eb1a5d6987  
2      ec6a82e4a1b9b4dc467639eb1a5d6987  
3      ec6a82e4a1b9b4dc467639eb1a5d6987  
4      ec6a82e4a1b9b4dc467639eb1a5d6987  
...                                 ...  
11244  7caa7e3d9378310a86ce7e20fc58b561  
11245  7caa7e3d9378310a86ce7e20fc58b561  
11246  7caa7e3d9378310a86ce7e20fc58b561  
11247  7caa7e3d9378310a86ce7e20fc58b561  
11248  7caa7e3d9378310a86ce7e20fc58b561  

[11249 rows x 7 columns]

Проверка на совпадение датафреймов:

In [24]:
print(event_name_12.equals(event_name_13))
print(event_name_12.equals(event_name_14))

True
True


Слияние датафреймов events в один c последующим присоединением таблицы event_name:

In [25]:
dfs = [events_12,events_13,events_14]
data = pd.DataFrame()
for df in dfs:
    data = pd.concat([data,df],ignore_index=True)
display(data)

id                         forecastMarker  \
0          1               SelfService/RUS/turnover   
1          2               SelfService/RUS/turnover   
2          3               SelfService/RUS/turnover   
3          4               SelfService/RUS/turnover   
4          5               SelfService/RUS/turnover   
...      ...                                    ...   
44726  11245  weekly_report.SS:new_users_90d:tiktok   
44727  11246  weekly_report.SS:new_users_90d:tiktok   
44728  11247  weekly_report.SS:new_users_90d:tiktok   
44729  11248  weekly_report.SS:new_users_90d:tiktok   
44730  11249  weekly_report.SS:new_users_90d:tiktok   

                     request_time  event_id            sender  \
0      2022-09-12 11:30:11.106628         1       test_system   
1      2022-09-12 11:30:11.111359        29       test_system   
2      2022-09-12 11:30:11.115989         2       test_system   
3      2022-09-12 11:30:12.761754         3       test_system   
4      2022-09-12 11:30:12.770134         5       test_system   
...                           ...       ...               ...   
44726  2022-09-14 06:07:47.582138        24  weekly_report.ss   
44727  2022-09-14 06:07:47.588185        15  weekly_report.ss   
44728  2022-09-14 06:07:47.594110        16  weekly_report.ss   
44729  2022-09-14 06:07:47.601623        17  weekly_report.ss   
44730  2022-09-14 06:07:47.609898        31  weekly_report.ss   

                                                 context  \
0      df = <class 'NoneType'>, sql_query = <class 'N...   
1                                                   None   
2                                                   None   
3                          Найдено 1 строк с параметрами   
4                                                   None   
...                                                  ...   
44726                                               None   
44727                                               None   
44728                                               None   
44729                                               None   
44730                         Отдан датафрейм: (1352, 4)   

                             session_id  
0      6f4277f6df57c4f4e90779966522c876  
1      6f4277f6df57c4f4e90779966522c876  
2      6f4277f6df57c4f4e90779966522c876  
3      6f4277f6df57c4f4e90779966522c876  
4      6f4277f6df57c4f4e90779966522c876  
...                                 ...  
44726  7caa7e3d9378310a86ce7e20fc58b561  
44727  7caa7e3d9378310a86ce7e20fc58b561  
44728  7caa7e3d9378310a86ce7e20fc58b561  
44729  7caa7e3d9378310a86ce7e20fc58b561  
44730  7caa7e3d9378310a86ce7e20fc58b561  

[44731 rows x 7 columns]

In [26]:
data = pd.merge(data, event_name_12, how='left', left_on = 'event_id', right_on = 'id').rename(columns = {'id_x':'id'}).drop('id_y',axis =1)    


In [27]:
display(data)

id                         forecastMarker  \
0          1               SelfService/RUS/turnover   
1          2               SelfService/RUS/turnover   
2          3               SelfService/RUS/turnover   
3          4               SelfService/RUS/turnover   
4          5               SelfService/RUS/turnover   
...      ...                                    ...   
44726  11245  weekly_report.SS:new_users_90d:tiktok   
44727  11246  weekly_report.SS:new_users_90d:tiktok   
44728  11247  weekly_report.SS:new_users_90d:tiktok   
44729  11248  weekly_report.SS:new_users_90d:tiktok   
44730  11249  weekly_report.SS:new_users_90d:tiktok   

                     request_time  event_id            sender  \
0      2022-09-12 11:30:11.106628         1       test_system   
1      2022-09-12 11:30:11.111359        29       test_system   
2      2022-09-12 11:30:11.115989         2       test_system   
3      2022-09-12 11:30:12.761754         3       test_system   
4      2022-09-12 11:30:12.770134         5       test_system   
...                           ...       ...               ...   
44726  2022-09-14 06:07:47.582138        24  weekly_report.ss   
44727  2022-09-14 06:07:47.588185        15  weekly_report.ss   
44728  2022-09-14 06:07:47.594110        16  weekly_report.ss   
44729  2022-09-14 06:07:47.601623        17  weekly_report.ss   
44730  2022-09-14 06:07:47.609898        31  weekly_report.ss   

                                                 context  \
0      df = <class 'NoneType'>, sql_query = <class 'N...   
1                                                   None   
2                                                   None   
3                          Найдено 1 строк с параметрами   
4                                                   None   
...                                                  ...   
44726                                               None   
44727                                               None   
44728                                               None   
44729                                               None   
44730                         Отдан датафрейм: (1352, 4)   

                             session_id  \
0      6f4277f6df57c4f4e90779966522c876   
1      6f4277f6df57c4f4e90779966522c876   
2      6f4277f6df57c4f4e90779966522c876   
3      6f4277f6df57c4f4e90779966522c876   
4      6f4277f6df57c4f4e90779966522c876   
...                                 ...   
44726  7caa7e3d9378310a86ce7e20fc58b561   
44727  7caa7e3d9378310a86ce7e20fc58b561   
44728  7caa7e3d9378310a86ce7e20fc58b561   
44729  7caa7e3d9378310a86ce7e20fc58b561   
44730  7caa7e3d9378310a86ce7e20fc58b561   

                                              event_name  
0                         Инициализируем свойства класса  
1          Начинаем получать гипермараметры и регрессоры  
2            Переход в функцию getParametrsAndRegressors  
3      Выход из функции getParametrsAndRegressors с п...  
4      Записали в свойства класса гиперпараметры и ре...  
...                                                  ...  
44726                saveForecastToCSV: выход из функции  
44727                     get_forecast: выход из функции  
44728               updateForecastAnswer: вход в функцию  
44729             updateForecastAnswer: выход из функции  
44730                             JSON отправлен обратно  

[44731 rows x 8 columns]

Пропусков в колонках нет:

In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44731 entries, 0 to 44730
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              44731 non-null  int64 
 1   forecastMarker  44731 non-null  object
 2   request_time    44731 non-null  object
 3   event_id        44731 non-null  int64 
 4   sender          44731 non-null  object
 5   context         44731 non-null  object
 6   session_id      44731 non-null  object
 7   event_name      44731 non-null  object
dtypes: int64(2), object(6)
memory usage: 3.1+ MB


###  Сохранение данных  <a class="anchor" id="id_3"></a>

In [29]:
data.to_csv( r'Elama\elama.csv', index=True) 

### Загрузка объединенного файла с данными за три дня <a class="anchor" id="id_4"></a>

In [30]:
engine = create_engine(f'sqlite:///{merged_file}', echo = False)
merged_base = pd.read_sql(query, engine)


Вывод полной базы:

In [40]:
query = """
WITH 
tmp AS (   -- the table is responsible for converting timestamp string from text type into real type
SELECT *,
       unixepoch(request_time)+cast (substr(request_time,20,7) as real) as request_time_unixepoch     
       
FROM events
),
-- Using Lag() function to find delta time between two consequent events in a session
tmp_1 AS ( SELECT *, 
           lag(request_time_unixepoch,1) OVER  (partition by session_id order by request_time_unixepoch) as previous_time
           FROM tmp
         ),
tmp_2 AS (SELECT id,
                 forecastMarker,
                 request_time,
                 CASE
                   WHEN event_id = 1 THEN 0.00
                   ELSE request_time_unixepoch - previous_time 
                 END as delta_by_session,
                 event_id,
                 sender,
                 context,
                 session_id                  
          FROM tmp_1
          ORDER BY request_time
         ),
         -- a service table responsible for checking NULL values
num_nulls AS (SELECT COUNT (session_id) as num_null_sessions,
                     COUNT (forecastMArker) as num_null_FM,
                     COUNT (context) as num_null_context,
                     COUNT (request_time) as num_null_reqTime,
                     COUNT (sender) as num_null_sender,
                     COUNT (id) as num_null_id
     
            FROM tmp_1
            WHERE session_id IS NULL 
                  OR forecastMarker is NULL
                  OR context IS NULL
                  OR request_time IS NULL
                  OR sender is NULL
                  OR id is NULL
              ),
events_by_desc_delta AS (SELECT forecastMarker,
                                request_time,
                                delta_by_session,
                                event_id,
                                event_name,
                                sender,
                                context,
                                session_id       
       
                         FROM ( SELECT forecastMarker,
                                       request_time,      
                                       delta_by_session,      
                                       rank () over(partition by event_id order by delta_by_session desc) as rank,
                                       event_id,
                                       event_name,
                                       sender,
                                       context,
                                       session_id
                                FROM tmp_2
                                JOIN event_name ON tmp_2.event_id = event_name.id
                                )
                         WHERE rank = 1 AND delta_by_session IS NOT NULL
                         ORDER BY delta_by_session desc
                         ),
                        
statistics AS (
       SELECT event_id,
              SUM((delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2))*
              (delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2)) ) / (COUNT(delta_by_session)-1)as variance, 
              sender,
              event_name
       FROM tmp_2
       JOIN event_name ON tmp_2.event_id = event_name.id
       group by event_id
       order by variance desc
       ),
marker_avg_duration AS (
                  SELECT forecastMarker, 
                         avg(delta_by_session) as avg_delta
                  FROM tmp_2 
                  GROUP BY session_id,forecastMarker
                  ORDER BY avg_delta desc
                  LIMIT  5
                  ),
time_intersections AS (
                      SELECT t1.id,
                            t2.id as next_id,
                            t1.delta_by_session,
                            t1.event_id,
                            t2.event_id as next_event_id,
                            t1.request_time,
                            t2.request_time as next_req_time,
                            t1.session_id,
                            t1.forecastMarker
                     FROM tmp_2 as t1
                     JOIN tmp_2 as t2 ON t1.session_id = t2.session_id                      
                     WHERE t2.id = t1.id + 1 AND  strftime('%d', t2.request_time) = strftime('%d', t1.request_time)  
                     ORDEr BY t1.request_time
                    
                      ),
        
time_intersections_2 AS (SELECT *,
                  request_time as event,
                 +1 AS counter
          FROM time_intersections
          UNION ALL
          SELECT   *,
                   next_req_time as event, 
                   -1 AS counter
          FROM time_intersections),
parallel_tasks AS (SELECT forecastMarker,
                          running_processes,
                          date(request_time) as date
                   FROM (
                         SELECT *,
                                SUM(SUM(counter))
                                OVER (partition by strftime('%d', request_time) ORDER BY event) AS running_processes
                         FROM time_intersections_2 
                         GROUP BY event)
                  ORDER BY 2 desc
                  LIMIT 10),
failed_tasks AS (
                    SELECT CASE forecastMarker
                               WHEN NULL THEN 'none'
                               ELSE COUNT(forecastMarker)
                               END as failed_tasks
                    FROM tmp_2 t1
                    WHERE NOT EXISTS (SELECT *
                                      FROM tmp_2 t2
                                      WHERE t1.forecastMarker = t2.forecastMarker 
                                            AND t2.event_id = 31
                                      )
                    
                   )
SELECT tmp_2.id,
       forecastMarker,
       request_time,
       delta_by_session,
       event_id,
       sender,
       context,
       session_id,
       event_name
       
FROM tmp_2
JOIN event_name ON tmp_2.event_id = event_name.id
ORDER BY request_time

"""

output_1 = pd.read_sql(query, engine)
display(output_1)

id                         forecastMarker  \
0          1               SelfService/RUS/turnover   
1          2               SelfService/RUS/turnover   
2          3               SelfService/RUS/turnover   
3          4               SelfService/RUS/turnover   
4          5               SelfService/RUS/turnover   
...      ...                                    ...   
44726  11245  weekly_report.SS:new_users_90d:tiktok   
44727  11246  weekly_report.SS:new_users_90d:tiktok   
44728  11247  weekly_report.SS:new_users_90d:tiktok   
44729  11248  weekly_report.SS:new_users_90d:tiktok   
44730  11249  weekly_report.SS:new_users_90d:tiktok   

                     request_time  delta_by_session  event_id  \
0      2022-09-12 11:30:11.106628          0.000000         1   
1      2022-09-12 11:30:11.111359          0.004731        29   
2      2022-09-12 11:30:11.115989          0.004630         2   
3      2022-09-12 11:30:12.761754          1.645765         3   
4      2022-09-12 11:30:12.770134          0.008380         5   
...                           ...               ...       ...   
44726  2022-09-14 06:07:47.582138          0.053248        24   
44727  2022-09-14 06:07:47.588185          0.006047        15   
44728  2022-09-14 06:07:47.594110          0.005925        16   
44729  2022-09-14 06:07:47.601623          0.007513        17   
44730  2022-09-14 06:07:47.609898          0.008275        31   

                 sender                                            context  \
0           test_system  df = <class 'NoneType'>, sql_query = <class 'N...   
1           test_system                                               None   
2           test_system                                               None   
3           test_system                      Найдено 1 строк с параметрами   
4           test_system                                               None   
...                 ...                                                ...   
44726  weekly_report.ss                                               None   
44727  weekly_report.ss                                               None   
44728  weekly_report.ss                                               None   
44729  weekly_report.ss                                               None   
44730  weekly_report.ss                         Отдан датафрейм: (1352, 4)   

                             session_id  \
0      6f4277f6df57c4f4e90779966522c876   
1      6f4277f6df57c4f4e90779966522c876   
2      6f4277f6df57c4f4e90779966522c876   
3      6f4277f6df57c4f4e90779966522c876   
4      6f4277f6df57c4f4e90779966522c876   
...                                 ...   
44726  7caa7e3d9378310a86ce7e20fc58b561   
44727  7caa7e3d9378310a86ce7e20fc58b561   
44728  7caa7e3d9378310a86ce7e20fc58b561   
44729  7caa7e3d9378310a86ce7e20fc58b561   
44730  7caa7e3d9378310a86ce7e20fc58b561   

                                              event_name  
0                         Инициализируем свойства класса  
1          Начинаем получать гипермараметры и регрессоры  
2            Переход в функцию getParametrsAndRegressors  
3      Выход из функции getParametrsAndRegressors с п...  
4      Записали в свойства класса гиперпараметры и ре...  
...                                                  ...  
44726                saveForecastToCSV: выход из функции  
44727                     get_forecast: выход из функции  
44728               updateForecastAnswer: вход в функцию  
44729             updateForecastAnswer: выход из функции  
44730                             JSON отправлен обратно  

[44731 rows x 9 columns]

Подсчет количества пропущенных значений в полях:

In [32]:
query = """
WITH 
tmp AS (
SELECT *,
       unixepoch(request_time)+cast (substr(request_time,20,7) as real) as request_time_unixepoch     
       
FROM events
),
tmp_1 AS ( SELECT *,
           lag(request_time_unixepoch,1) OVER  (partition by session_id order by request_time_unixepoch) as previous_time
           FROM tmp
         ),
tmp_2 AS (SELECT id,
                 forecastMarker,
                 request_time,
                 CASE
                   WHEN event_id = 1 THEN 0.00
                   ELSE request_time_unixepoch - previous_time 
                 END as delta_by_session,
                 event_id,
                 sender,
                 context,
                 session_id                  
          FROM tmp_1
          ORDER BY request_time
         ),
num_nulls AS (SELECT COUNT (session_id) as num_null_sessions,
                     COUNT (forecastMArker) as num_null_FM,
                     COUNT (context) as num_null_context,
                     COUNT (request_time) as num_null_reqTime,
                     COUNT (sender) as num_null_sender,
                     COUNT (id) as num_null_id
     
            FROM tmp_1
            WHERE session_id IS NULL 
                  OR forecastMarker is NULL
                  OR context IS NULL
                  OR request_time IS NULL
                  OR sender is NULL
                  OR id is NULL
              ),
events_by_desc_delta AS (SELECT forecastMarker,
                                request_time,
                                delta_by_session,
                                event_id,
                                event_name,
                                sender,
                                context,
                                session_id       
       
                         FROM ( SELECT forecastMarker,
                                       request_time,      
                                       delta_by_session,      
                                       rank () over(partition by event_id order by delta_by_session desc) as rank,
                                       event_id,
                                       event_name,
                                       sender,
                                       context,
                                       session_id
                                FROM tmp_2
                                JOIN event_name ON tmp_2.event_id = event_name.id
                                )
                         WHERE rank = 1 AND delta_by_session IS NOT NULL
                         ORDER BY delta_by_session desc
                         ),
                        
statistics AS (
       SELECT event_id,
              SUM((delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2))*
              (delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2)) ) / (COUNT(delta_by_session)-1)as variance, 
              sender,
              event_name
       FROM tmp_2
       JOIN event_name ON tmp_2.event_id = event_name.id
       group by event_id
       order by variance desc
       ),
marker_avg_duration AS (
                  SELECT forecastMarker, 
                         avg(delta_by_session) as avg_delta
                  FROM tmp_2 
                  GROUP BY session_id,forecastMarker
                  ORDER BY avg_delta desc
                  LIMIT  5
                  ),
time_intersections AS (
                      SELECT t1.id,
                            t2.id as next_id,
                            t1.delta_by_session,
                            t1.event_id,
                            t2.event_id as next_event_id,
                            t1.request_time,
                            t2.request_time as next_req_time,
                            t1.session_id,
                            t1.forecastMarker
                     FROM tmp_2 as t1
                     JOIN tmp_2 as t2 ON t1.session_id = t2.session_id                      
                     WHERE t2.id = t1.id + 1 AND  strftime('%d', t2.request_time) = strftime('%d', t1.request_time)  
                     ORDEr BY t1.request_time
                    
                      ),
        
time_intersections_2 AS (SELECT *,
                  request_time as event,
                 +1 AS counter
          FROM time_intersections
          UNION ALL
          SELECT   *,
                   next_req_time as event, 
                   -1 AS counter
          FROM time_intersections),
parallel_tasks AS (SELECT forecastMarker,
                          running_processes,
                          date(request_time) as date
                   FROM (
                         SELECT *,
                                SUM(SUM(counter))
                                OVER (partition by strftime('%d', request_time) ORDER BY event) AS running_processes
                         FROM time_intersections_2 
                         GROUP BY event)
                  ORDER BY 2 desc
                  LIMIT 10),
failed_tasks AS (
                    SELECT CASE forecastMarker
                               WHEN NULL THEN 'none'
                               ELSE COUNT(forecastMarker)
                               END as failed_tasks
                    FROM tmp_2 t1
                    WHERE NOT EXISTS (SELECT *
                                      FROM tmp_2 t2
                                      WHERE t1.forecastMarker = t2.forecastMarker 
                                            AND t2.event_id = 31
                                      )
                    
                   )
SELECT *
FROM num_nulls

"""

output_2 = pd.read_sql(query, engine)
display(output_2)

num_null_sessions  num_null_FM  num_null_context  num_null_reqTime  \
0                  0            0                 0                 0   

   num_null_sender  num_null_id  
0                0            0

##  Ответы на вопросы   <a class="anchor" id="id_5"></a>

Важно отметить: 
Первое событие для каждой сессии имеет идентификатор  1.
1. Идентификаторы
последующих событий не соответствуют последовательности наступления.
2. Шаг событий - это интервал времени между событиями наступающими друг за
другом.  
3. Задача прогнозирования считается завершенной успешно, если наступило
событие с идентификатором 31.  
4. Некоторые задачи прогнозирования могут использовать прогнозы по типу
рекурсии. Т.е. при построении прогноза запрашивать прогноз другого
параметра, как регрессора. Например при построении прогноза активаций
пользователей запрашивается подпрогноз с количеством регистраций. Т.к.
регистрации напрямую влияют на количество активаций.  

###   1 Между какими событиями наибольший шаг? Укажите пару идентификаторов  событий с наибольшим шагом, относительно всей базы (всех файлов). <a class="anchor" id="id_6"></a>

Самый долгий шаг между событием 7 и 6 в следующей сессии:

In [33]:
query = """
WITH 
tmp AS (
SELECT *,
       unixepoch(request_time)+cast (substr(request_time,20,7) as real) as request_time_unixepoch     
       
FROM events
),
tmp_1 AS ( SELECT *,
             lag(request_time_unixepoch,1) OVER  (partition by session_id order by request_time_unixepoch) as previous_time
           FROM tmp
         ),
tmp_2 AS (SELECT id,
                 forecastMarker,
                 request_time,
                 CASE
                   WHEN event_id = 1 THEN 0.00
                   ELSE request_time_unixepoch - previous_time 
                 END as delta_by_session,
                 event_id,
                 sender,
                 context,
                 session_id                
                  
          FROM tmp_1
          ORDER BY request_time
         ),
num_nulls AS (SELECT COUNT (session_id) as num_null_sessions,
                     COUNT (forecastMArker) as num_null_FM,
                     COUNT (context) as num_null_context,
                     COUNT (request_time) as num_null_reqTime,
                     COUNT (sender) as num_null_sender,
                     COUNT (id) as num_null_id
     
            FROM tmp_1
            WHERE session_id IS NULL 
                  OR forecastMarker is NULL
                  OR context IS NULL
                  OR request_time IS NULL
                  OR sender is NULL
                  OR id is NULL
              )

SELECT DISTINCT t1.id,
       t2.id as previous_id,
       t1.delta_by_session,
       t1.event_id,
       t2.event_id as previous_event_id,
       t1.request_time,
       t2.request_time as previous_req_time,
       t1.session_id,
       en.event_name,
       en2.event_name as prev_event_name
FROM tmp_2 as t1
JOIN tmp_2 as t2 ON t1.session_id = t2.session_id
JOIN event_name en ON t1.event_id = en.id
JOIN event_name en2 ON t2.event_id = en2.id
WHERE t2.id = t1.id - 1  
ORDER by t1.delta_by_session desc
LIMIT 1
                  
"""

output_3 = pd.read_sql(query, engine)
display(output_3)

id  previous_id  delta_by_session  event_id  previous_event_id  \
0  5744         5743         106.54821         7                  6   

                 request_time           previous_req_time  \
0  2022-09-14 05:22:40.308128  2022-09-14 05:20:53.759918   

                         session_id  \
0  4dcab2af4573ced8251fb7433174ba15   

                                         event_name  \
0  Получили данные из BQ по полученному SQL запросу   

                                     prev_event_name  
0  Начинаем читать данные из BQ по полученному SQ...

Топ 5 уникальных событий по времени их выполнения:

In [34]:
query = """
WITH 
tmp AS (
SELECT *,
       unixepoch(request_time)+cast (substr(request_time,20,7) as real) as request_time_unixepoch     
       
FROM events
),
tmp_1 AS ( SELECT *,
           lag(request_time_unixepoch,1) OVER  (partition by session_id order by request_time_unixepoch) as previous_time
           FROM tmp
         ),
tmp_2 AS (SELECT id,
                 forecastMarker,
                 request_time,
                 CASE
                   WHEN event_id = 1 THEN 0.00
                   ELSE request_time_unixepoch - previous_time 
                 END as delta_by_session,
                 event_id,
                 sender,
                 context,
                 session_id                
                  
          FROM tmp_1
          ORDER BY request_time
         ),
num_nulls AS (SELECT COUNT (session_id) as num_null_sessions,
                     COUNT (forecastMArker) as num_null_FM,
                     COUNT (context) as num_null_context,
                     COUNT (request_time) as num_null_reqTime,
                     COUNT (sender) as num_null_sender,
                     COUNT (id) as num_null_id
     
            FROM tmp_1
            WHERE session_id IS NULL 
                  OR forecastMarker is NULL
                  OR context IS NULL
                  OR request_time IS NULL
                  OR sender is NULL
                  OR id is NULL
              ),
events_by_desc_delta AS (SELECT forecastMarker,
                                request_time,
                                delta_by_session,
                                event_id,
                                event_name,
                                sender,
                                context,
                                session_id       
       
                         FROM ( SELECT forecastMarker,
                                       request_time,      
                                       delta_by_session,      
                                       rank () over(partition by event_id order by delta_by_session desc) as rank,
                                       event_id,
                                       event_name,
                                       sender,
                                       context,
                                       session_id
                                FROM tmp_2
                                JOIN event_name ON tmp_2.event_id = event_name.id)
                                WHERE rank = 1 AND delta_by_session IS NOT NULL
                                ORDER BY delta_by_session desc
                         )

SELECT *
FROM events_by_desc_delta
LIMIT 5
"""

output_4 = pd.read_sql(query, engine)
display(output_4)

forecastMarker                request_time  \
0  weekly_report.care:count_of_jivo_SA  2022-09-14 05:22:40.308128   
1             SelfService/RUS/turnover  2022-09-14 04:50:28.564784   
2                 SubAgency/RUS/ag_new  2022-09-13 04:53:51.247012   
3                                 None  2022-09-14 04:50:29.662821   
4          weekly_report.SAAS:visitors  2022-09-12 17:09:19.655564   

   delta_by_session  event_id  \
0        106.548210         7   
1         26.146893        28   
2         20.012902        21   
3         17.756815        23   
4         15.668156        34   

                                          event_name  \
0   Получили данные из BQ по полученному SQL запросу   
1     getForecastMarkerAsRegressor: выход из функции   
2                get_error_for_model: вход в функцию   
3                  saveForecastToCSV: вход в функцию   
4  get_error_for_model: получили performance_metrics   

                                        sender  \
0                           weekly_report.care   
1            DAG : bq_analytics_tables_refresh   
2  forecast_for_PHR.load_prophet_forecast_data   
3                  bq_analytics_tables_refresh   
4                           weekly_report.saas   

                                          context  \
0  Получили из BQ набор данных размером: (548, 2)   
1                                            None   
2                                            None   
3                                            None   
4                                            None   

                         session_id  
0  4dcab2af4573ced8251fb7433174ba15  
1  263f9c139d1c9aec358b5f4cadd85aee  
2  c972050a7007eb6d518fff16f5ad0abd  
3  d14e4990e3162d98a3ed61ee30005cf6  
4  f151aa3db81db96c276313c910abca75

###  2 Интервал времени до наступления какого события показывает наибольший разброс. Укажите пару идентификатор и sender с наибольшим разбросом. <a class="anchor" id="id_7"></a>

In [35]:
query = """
WITH 
tmp AS (
SELECT *,
       unixepoch(request_time)+cast (substr(request_time,20,7) as real) as request_time_unixepoch     
       
FROM events
),
tmp_1 AS ( SELECT *,
           lag(request_time_unixepoch,1) OVER  (partition by session_id order by request_time_unixepoch) as previous_time
           FROM tmp
         ),
tmp_2 AS (SELECT id,
                 forecastMarker,
                 request_time,
                 CASE
                   WHEN event_id = 1 THEN 0.00
                   ELSE request_time_unixepoch - previous_time 
                 END as delta_by_session,
                 event_id,
                 sender,
                 context,
                 session_id                
                  
          FROM tmp_1
          ORDER BY request_time
         ),
num_nulls AS (SELECT COUNT (session_id) as num_null_sessions,
                     COUNT (forecastMArker) as num_null_FM,
                     COUNT (context) as num_null_context,
                     COUNT (request_time) as num_null_reqTime,
                     COUNT (sender) as num_null_sender,
                     COUNT (id) as num_null_id
     
            FROM tmp_1
            WHERE session_id IS NULL 
                  OR forecastMarker is NULL
                  OR context IS NULL
                  OR request_time IS NULL
                  OR sender is NULL
                  OR id is NULL
              ),
events_by_desc_delta AS (SELECT request_time,
                                delta_by_session,
                                event_id,
                                event_name,
                                sender,
                                context,
                                session_id       
       
                         FROM ( SELECT request_time,      
                                       delta_by_session,      
                                       rank () over(partition by event_id order by delta_by_session desc) as rank,
                                       event_id,
                                       event_name,
                                       sender,
                                       context,
                                       session_id
                                FROM tmp_2
                                JOIN event_name ON tmp_2.event_id = event_name.id
                                )
                         WHERE rank = 1 AND delta_by_session IS NOT NULL
                         ORDER BY delta_by_session desc
                         ),

statistics AS (
       SELECT event_id,
              SUM((delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2))*
              (delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2)) ) / (COUNT(delta_by_session)-1)as variance, 
              sender,
              event_name
       FROM tmp_2
       JOIN event_name ON tmp_2.event_id = event_name.id
       group by event_id
       order by variance desc
       )
SELECT *
FROM statistics
LIMIT 5


"""

output_5 = pd.read_sql(query, engine)
display(output_5)

event_id    variance                                             sender  \
0        28  258.739251                                        test_system   
1         7   63.789895                                        test_system   
2        21    9.115613  internal_call for SelfService/RUS/turnover, se...   
3        23    5.308505  internal_call for SelfService/RUS/turnover, se...   
4        34    4.860770  internal_call for SelfService/RUS/turnover, se...   

                                          event_name  
0     getForecastMarkerAsRegressor: выход из функции  
1   Получили данные из BQ по полученному SQL запросу  
2                get_error_for_model: вход в функцию  
3                  saveForecastToCSV: вход в функцию  
4  get_error_for_model: получили performance_metrics

Наибольший разброс имеет пара sender: test_system - event_name:getForecastMarkerAsRegressor: выход из функции с показателем дисперсии 258.739251 секунд. По моему мнению, это связано с изменениями различных параметров системы при тестовых запусках, что сказывается на скорости выполнения. 

### 3 Какая задача прогнозирования выполняется дольше всего (название задачи)? <a class="anchor" id="id_8"></a>

In [36]:
query = """
WITH 
tmp AS (
SELECT *,
       unixepoch(request_time)+cast (substr(request_time,20,7) as real) as request_time_unixepoch     
       
FROM events
),
tmp_1 AS ( SELECT *,
           lag(request_time_unixepoch,1) OVER  (partition by session_id order by request_time_unixepoch) as previous_time
           FROM tmp
         ),
tmp_2 AS (SELECT id,
                 forecastMarker,
                 request_time,
                 CASE
                   WHEN event_id = 1 THEN 0.00
                   ELSE request_time_unixepoch - previous_time 
                 END as delta_by_session,
                 event_id,
                 sender,
                 context,
                 session_id                
                  
          FROM tmp_1
          ORDER BY request_time
         ),
num_nulls AS (SELECT COUNT (session_id) as num_null_sessions,
                     COUNT (forecastMArker) as num_null_FM,
                     COUNT (context) as num_null_context,
                     COUNT (request_time) as num_null_reqTime,
                     COUNT (sender) as num_null_sender,
                     COUNT (id) as num_null_id
     
            FROM tmp_1
            WHERE session_id IS NULL 
                  OR forecastMarker is NULL
                  OR context IS NULL
                  OR request_time IS NULL
                  OR sender is NULL
                  OR id is NULL
              ),
events_by_desc_delta AS (SELECT forecastMarker,
                                request_time,
                                delta_by_session,
                                event_id,
                                event_name,
                                sender,
                                context,
                                session_id       
       
                         FROM ( SELECT forecastMarker,
                                       request_time,      
                                       delta_by_session,      
                                       rank () over(partition by event_id order by delta_by_session desc) as rank,
                                       event_id,
                                       event_name,
                                       sender,
                                       context,
                                       session_id
                                FROM tmp_2
                                JOIN event_name ON tmp_2.event_id = event_name.id
                                )
                         WHERE rank = 1 AND delta_by_session IS NOT NULL
                         ORDER BY delta_by_session desc
                         ),
                        
statistics AS (
       SELECT event_id,
              SUM((delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2))*
              (delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2)) ) / (COUNT(delta_by_session)-1)as variance, 
              sender,
              event_name
       FROM tmp_2
       JOIN event_name ON tmp_2.event_id = event_name.id
       group by event_id
       order by variance desc),
marker_avg_duration AS(
                  SELECT forecastMarker, 
                         avg(delta_by_session) as avg_delta
                  FROM tmp_2 
                  GROUP BY session_id
                  ORDER BY avg_delta desc
                  LIMIT  5
                  )

SELECT *
FROM marker_avg_duration

"""

output_6 = pd.read_sql(query, engine)
display(output_6)

forecastMarker  avg_delta
0  weekly_report.care:count_of_jivo_SA   4.326326
1     weekly_report.care:sum_of_income   2.790707
2  weekly_report.care:count_of_jivo_SS   2.697098
3          weekly_report.SAAS:visitors   2.450200
4          weekly_report.SAAS:visitors   2.377289

Задача weekly_report.care:count_of_jivo_SA	обычно выполняется больше всего: 4.326326 секунды.

### 4 Какое количество задач прогнозирования могут выполняться одновременно?      Укажите максимальное число параллельных задач. <a class="anchor" id="id_9"></a>

In [37]:
query = """
WITH 
tmp AS (
SELECT *,
       unixepoch(request_time)+cast (substr(request_time,20,7) as real) as request_time_unixepoch     
       
FROM events
),
tmp_1 AS ( SELECT *,
           lag(request_time_unixepoch,1) OVER  (partition by session_id order by request_time_unixepoch) as previous_time
           FROM tmp
         ),
tmp_2 AS (SELECT id,
                 forecastMarker,
                 request_time,
                 CASE
                   WHEN event_id = 1 THEN 0.00
                   ELSE request_time_unixepoch - previous_time 
                 END as delta_by_session,
                 event_id,
                 sender,
                 context,
                 session_id                  
          FROM tmp_1
          ORDER BY request_time
         ),
num_nulls AS (SELECT COUNT (session_id) as num_null_sessions,
                     COUNT (forecastMArker) as num_null_FM,
                     COUNT (context) as num_null_context,
                     COUNT (request_time) as num_null_reqTime,
                     COUNT (sender) as num_null_sender,
                     COUNT (id) as num_null_id
     
            FROM tmp_1
            WHERE session_id IS NULL 
                  OR forecastMarker is NULL
                  OR context IS NULL
                  OR request_time IS NULL
                  OR sender is NULL
                  OR id is NULL
              ),
events_by_desc_delta AS (SELECT forecastMarker,
                                request_time,
                                delta_by_session,
                                event_id,
                                event_name,
                                sender,
                                context,
                                session_id       
       
                         FROM ( SELECT forecastMarker,
                                       request_time,      
                                       delta_by_session,      
                                       rank () over(partition by event_id order by delta_by_session desc) as rank,
                                       event_id,
                                       event_name,
                                       sender,
                                       context,
                                       session_id
                                FROM tmp_2
                                JOIN event_name ON tmp_2.event_id = event_name.id
                                )
                         WHERE rank = 1 AND delta_by_session IS NOT NULL
                         ORDER BY delta_by_session desc
                         ),
                        
statistics AS (
       SELECT event_id,
              SUM((delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2))*
              (delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2)) ) / (COUNT(delta_by_session)-1)as variance, 
              sender,
              event_name
       FROM tmp_2
       JOIN event_name ON tmp_2.event_id = event_name.id
       group by event_id
       order by variance desc),
marker_avg_duration AS (
                  SELECT forecastMarker, 
                         avg(delta_by_session) as avg_delta
                  FROM tmp_2 
                  GROUP BY session_id,forecastMarker
                  ORDER BY avg_delta desc
                  LIMIT  5
                  ),
time_intersections AS (
                      SELECT t1.id,
                            t2.id as next_id,
                            t1.delta_by_session,
                            t1.event_id,
                            t2.event_id as next_event_id,
                            t1.request_time,
                            t2.request_time as next_req_time,
                            t1.session_id,
                            t1.forecastMarker
                     FROM tmp_2 as t1
                     JOIN tmp_2 as t2 ON t1.session_id = t2.session_id                      
                     WHERE t2.id = t1.id + 1 AND  strftime('%d', t2.request_time) = strftime('%d', t1.request_time)  
                     ORDEr BY t1.request_time
                    
                      ),
        
time_intersections_2 AS (SELECT *,
                  request_time as event,
                 +1 AS counter
          FROM time_intersections
          UNION ALL
          SELECT   *,
                   next_req_time as event, 
                   -1 AS counter
          FROM time_intersections),
parallel_tasks AS (SELECT forecastMarker,
                          running_processes,
                          date(request_time) as date
                   FROM (
                         SELECT *,
                                SUM(SUM(counter))
                                OVER (partition by strftime('%d', request_time) ORDER BY event) AS running_processes
                         FROM time_intersections_2 
                         GROUP BY event)
                  ORDER BY 2 desc
                  LIMIT 10)
SELECT *
FROM parallel_tasks
    


"""

output_7 = pd.read_sql(query, engine)
display(output_7)

forecastMarker  running_processes  \
0                                               None                  3   
1                                               None                  3   
2                        SaaS/RUS/turnover/vkontakte                  3   
3                            SelfService/RUS/ag_core                  3   
4  weekly_report.SAAS:new_visitors:utm_medium:cpc...                  2   
5  weekly_report.SAAS:new_visitors:utm_medium:cpc...                  2   
6  weekly_report.SAAS:new_visitors:utm_medium:cpc...                  2   
7  weekly_report.SAAS:new_visitors:utm_medium:mes...                  2   
8                              u.malyshenko@elama.ru                  2   
9                              u.malyshenko@elama.ru                  2   

         date  
0  2022-09-13  
1  2022-09-14  
2  2022-09-14  
3  2022-09-14  
4  2022-09-12  
5  2022-09-12  
6  2022-09-12  
7  2022-09-13  
8  2022-09-13  
9  2022-09-13

Максимум параллельно идущих прогнозов равно 3.

### 5  Перечислите forecastMarker которые не завершились успешно <a class="anchor" id="id_10"></a>

In [38]:
query = """
WITH 
tmp AS (
SELECT *,
       unixepoch(request_time)+cast (substr(request_time,20,7) as real) as request_time_unixepoch     
       
FROM events
),
tmp_1 AS ( SELECT *,
           lag(request_time_unixepoch,1) OVER  (partition by session_id order by request_time_unixepoch) as previous_time
           FROM tmp
         ),
tmp_2 AS (SELECT id,
                 forecastMarker,
                 request_time,
                 CASE
                   WHEN event_id = 1 THEN 0.00
                   ELSE request_time_unixepoch - previous_time 
                 END as delta_by_session,
                 event_id,
                 sender,
                 context,
                 session_id                  
          FROM tmp_1
          ORDER BY request_time
         ),
num_nulls AS (SELECT COUNT (session_id) as num_null_sessions,
                     COUNT (forecastMArker) as num_null_FM,
                     COUNT (context) as num_null_context,
                     COUNT (request_time) as num_null_reqTime,
                     COUNT (sender) as num_null_sender,
                     COUNT (id) as num_null_id
     
            FROM tmp_1
            WHERE session_id IS NULL 
                  OR forecastMarker is NULL
                  OR context IS NULL
                  OR request_time IS NULL
                  OR sender is NULL
                  OR id is NULL
              ),
events_by_desc_delta AS (SELECT forecastMarker,
                                request_time,
                                delta_by_session,
                                event_id,
                                event_name,
                                sender,
                                context,
                                session_id       
       
                         FROM ( SELECT forecastMarker,
                                       request_time,      
                                       delta_by_session,      
                                       rank () over(partition by event_id order by delta_by_session desc) as rank,
                                       event_id,
                                       event_name,
                                       sender,
                                       context,
                                       session_id
                                FROM tmp_2
                                JOIN event_name ON tmp_2.event_id = event_name.id
                                )
                         WHERE rank = 1 AND delta_by_session IS NOT NULL
                         ORDER BY delta_by_session desc
                         ),
                        
statistics AS (
       SELECT event_id,
              SUM((delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2))*
              (delta_by_session-(SELECT AVG(delta_by_session) FROM tmp_2)) ) / (COUNT(delta_by_session)-1)as variance, 
              sender,
              event_name
       FROM tmp_2
       JOIN event_name ON tmp_2.event_id = event_name.id
       group by event_id
       order by variance desc
       ),
marker_avg_duration AS (
                  SELECT forecastMarker, 
                         avg(delta_by_session) as avg_delta
                  FROM tmp_2 
                  GROUP BY session_id,forecastMarker
                  ORDER BY avg_delta desc
                  LIMIT  5
                  ),
time_intersections AS (
                      SELECT t1.id,
                            t2.id as next_id,
                            t1.delta_by_session,
                            t1.event_id,
                            t2.event_id as next_event_id,
                            t1.request_time,
                            t2.request_time as next_req_time,
                            t1.session_id,
                            t1.forecastMarker
                     FROM tmp_2 as t1
                     JOIN tmp_2 as t2 ON t1.session_id = t2.session_id                      
                     WHERE t2.id = t1.id + 1 AND  strftime('%d', t2.request_time) = strftime('%d', t1.request_time)  
                     ORDEr BY t1.request_time
                    
                      ),
        
time_intersections_2 AS (SELECT *,
                  request_time as event,
                 +1 AS counter
          FROM time_intersections
          UNION ALL
          SELECT   *,
                   next_req_time as event, 
                   -1 AS counter
          FROM time_intersections),
parallel_tasks AS (SELECT forecastMarker,
                          running_processes,
                          date(request_time) as date
                   FROM (
                         SELECT *,
                                SUM(SUM(counter))
                                OVER (partition by strftime('%d', request_time) ORDER BY event) AS running_processes
                         FROM time_intersections_2 
                         GROUP BY event)
                  ORDER BY 2 desc
                  LIMIT 10),
failed_tasks AS (
                    SELECT CASE forecastMarker
                               WHEN NULL THEN 'none'
                               ELSE COUNT(forecastMarker)
                               END as failed_tasks
                    FROM tmp_2 t1
                    WHERE NOT EXISTS (SELECT *
                                      FROM tmp_2 t2
                                      WHERE t1.forecastMarker = t2.forecastMarker 
                                            AND t2.event_id = 31
                                      )
                    
                   )


SELECT *
FROM failed_tasks
"""

output_8 = pd.read_sql(query, engine)
display(output_8)

failed_tasks
0             0

По моим подсчетам все прогнозы завершились успешно.

## Выводы <a class="anchor" id="id_11"></a>

1. Самый продолжительный шаг составил 106.54821	секунды между событиями с индентификаторами 7 и	6	с  датой 2022-09-14 05:22:40.308128	  
2. Наибольший разброс имеет пара sender: test_system - event_name:getForecastMarkerAsRegressor: выход из функции с показателем дисперсии 258.739251 секунд.   
3. Задача weekly_report.care:count_of_jivo_SA обычно выполняется больше всего: 4.326326 секунды.  
4. Максимум параллельно идущих прогнозов равно 3.   
5. Неосуществленных прогнозов не обнаружено.